rust/rbac-registration/src/cardano/cip509/utils/cip134/uri_set.rs

```diff
@@ -0,0 +1,316 @@
+                    if let Some(uris) = metadata.certificate_uris.x_uris().get(&index) {
+                        x_uris.insert(index, uris.clone());
+                    }
+                },
+            }
+        }
+
+        for (index, cert) in metadata.c509_certs.iter().enumerate() {
+            match cert {
+                C509Cert::Undefined => {
+                    // The certificate wasn't changed - there is nothing to do.
+                },
+                C509Cert::Deleted => {
+                    c_uris.remove(&index);
+                },
+                C509Cert::C509CertInMetadatumReference(_) => {
+                    debug!("Ignoring unsupported metadatum reference");
+                },
+                C509Cert::C509Certificate(_) => {
+                    if let Some(uris) = metadata.certificate_uris.c_uris().get(&index) {
+                        c_uris.insert(index, uris.clone());
+                    }
+                },
+            }
+        }
+
+        Self(Arc::new(Cip0134UriSetInner { x_uris, c_uris }))
+    }
+}
+
+/// Iterates over X509 certificates and extracts CIP-0134 URIs.
+fn extract_x509_uris(certificates: &[X509DerCert]) -> Result<UrisMap> {
+    let mut result = UrisMap::new();
+
+    for (index, cert) in certificates.iter().enumerate() {
+        let X509DerCert::X509Cert(cert) = cert else {
+            continue;
+        };
+        // Find the "subject alternative name" extension.
+        let Some(extension) = cert
+            .tbs_certificate
+            .extensions
+            .iter()
+            .flatten()
+            .find(|e| e.extn_id == ID_CE_SUBJECT_ALT_NAME)
+        else {
+            continue;
+        };
+        let (_, der) = parse_der_sequence(extension.extn_value.as_bytes()).with_context(|| {
+            format!("Failed to parse DER sequence for Subject Alternative Name ({extension:?})")
+        })?;
+
+        let mut uris = Vec::new();
+        for data in der.ref_iter() {
+            if data.header.raw_tag() != Some(&[URI]) {
+                continue;
+            }
+            let content = data
+                .content
+                .as_slice()
+                .with_context(|| "Unable to process content for {data:?}")?;
+            let address = match decode_utf8(content) {
+                Ok(a) => a,
+                Err(e) => {
+                    // X.509 doesn't restrict the "alternative name" extension to be utf8 only, so
+                    // we cannot treat this as error.
+                    debug!("Ignoring {e:?}");
+                    continue;
+                },
+            };
+            let uri = match Cip0134Uri::parse(&address) {
+                Ok(u) => u,
+                Err(e) => {
+                    // Same as above - simply skip non-confirming values.
+                    debug!("Ignoring invalid CIP-0134 address: {e:?}");
+                    continue;
+                },
+            };
```

Comment:
Feel like this should be a method of `Cip509URI` that tries to make it from a `slice` of bytes, rather than done inline here.